code/datums/tracking_sound.dm

```diff
@@ -0,0 +1,223 @@
+#define LISTENER_OFFSET_UNKNOWN -LISTENER_OFFSET_GAP
+/// Gets the listener offset for a listener.
+#define LISTENER_OFFSET_GET(listener) (listeners[listener] - LISTENER_OFFSET_GAP)
+/// Sets the listener offset for a listener.
+#define LISTENER_OFFSET_SET(listener, offset) (listeners[listener] = offset + LISTENER_OFFSET_GAP)
+
+/// A sound source that tracks listeners and the source of the sound.
+/datum/sound_spatial_tracker
+	// vars for playsound_local passthrough
+	var/atom/source
+	var/sound/sound // this one is only for restarting a sound if the client loses it
+	var/base_volume
+	var/falloff_exponent
+	var/channel
+	var/pressure_affected
+	var/max_distance
+	var/falloff_distance
+	var/use_reverb
+
+	/// world.timeofday we started playing the sound.
+	var/start_time
+
+	/// sound length
+	var/sound_length
+
+	/// The listeners of the sound.
+	var/list/mob/listeners = list()
+
+	/// Mobs who have left any cells we care about
+	var/list/mob/leavers = list()
+
+	/// The spatial tracker for the sound.
+	var/datum/cell_tracker/spatial_tracker
+
+	/// List of cells we're tracking
+	var/list/cells = list()
+
+	/// Set to true if we were able to track sound length for self deletion.
+	var/qdel_scheduled = FALSE
+
+	/// Map of client ckeys to their expected offset, offset
+	var/list/client_offsets = list()
+
+/datum/sound_spatial_tracker/New(
+	source,
+	sound,
+	base_volume,
+	falloff_exponent,
+	channel,
+	pressure_affected,
+	max_distance,
+	falloff_distance,
+	use_reverb,
+	sound_length,
+)
+	src.source = source
+	src.sound = sound(file = sound, channel = channel)
+	src.base_volume = base_volume
+	src.falloff_exponent = falloff_exponent
+	src.channel = channel
+	src.pressure_affected = pressure_affected
+	src.max_distance = max_distance
+	src.falloff_distance = falloff_distance
+	src.use_reverb = use_reverb
+
+	start_time = REALTIMEOFDAY
+	GLOB.sound_spatial_trackers[channel] = src
+	spatial_tracker = new(max_distance, max_distance)
+	update_spatial_tracker()
+	src.sound_length = sound_length
+	if(sound_length)
+		schedule_qdel(sound_length)
+	RegisterSignal(source, COMSIG_QDELETING, PROC_REF(source_gone))
+	return ..()
+
+/datum/sound_spatial_tracker/proc/schedule_qdel(length)
+	if(qdel_scheduled)
+		return
+	qdel_scheduled = QDEL_IN_STOPPABLE(src, length)
+
+/datum/sound_spatial_tracker/proc/source_gone()
+	SIGNAL_HANDLER
+	qdel(src)
+
+/datum/sound_spatial_tracker/Destroy(force, ...)
+	source = null
+	for(var/mob/listener as anything in listeners)
+		release_listener(listener)
+	listeners.Cut()
+	if(GLOB.sound_spatial_trackers[channel] == src)
+		GLOB.sound_spatial_trackers[channel] = null
+	spatial_tracker = null
+	cells.Cut()
+	return ..()
+
+/datum/sound_spatial_tracker/process(seconds_per_tick)
+	for(var/mob/listent as anything in leavers)
+		release_listener(listent)
+	leavers.Cut()
+
+/datum/sound_spatial_tracker/proc/on_source_moved()
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, PROC_REF(update_spatial_tracker))
```

Comment:
What's sleeping that you need this? Spatial tracker updates probably shouldn't be sleeping in the first place

Reply:
update_listener potentially sleeps due to soundquery to get offset

Reply:
ok idk bro